Fix #3506, #5950, #5951: Add support for remote platform parameters [Blocked: #6062]

BenHenning opened the pull request:
TODO: Finish one PR is actually code complete.

<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [ ] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [ ] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [ ] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [ ] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [ ] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
<!-- Delete these section if this PR does not include UI-related changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-A11y-Guide))
- For PRs introducing new UI elements or color changes, both light and dark mode screenshots must be included
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

BenHenning committed:
Add support for remote platform parameters.

This includes a bunch of troubleshooting and testing for workers which
have been found to be broken. Also, additional work is needed to make
the production side of remote parameter syncing work. However, syncing
does work with the backup migration server on developer builds.

BenHenning committed:
Merge branch 'fix-work-manager-and-app-init' into add-support-for-remote-platform-parameters

Conflicts:
	app/src/main/AndroidManifest.xml
	app/src/main/java/org/oppia/android/app/application/AbstractOppiaApplication.kt
	app/src/main/java/org/oppia/android/app/application/dev/DeveloperApplicationComponent.kt
	domain/src/main/java/org/oppia/android/domain/oppialogger/logscheduler/MetricLogSchedulingWorker.kt
	domain/src/main/java/org/oppia/android/domain/oppialogger/logscheduler/MetricLogSchedulingWorkerFactory.kt
	domain/src/main/java/org/oppia/android/domain/oppialogger/loguploader/LogReportWorkManagerInitializer.kt
	domain/src/main/java/org/oppia/android/domain/oppialogger/loguploader/LogUploadWorkerFactory.kt
	domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorkManagerInitializer.kt
	domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorkerFactory.kt
	domain/src/main/java/org/oppia/android/domain/workmanager/WorkManagerConfigurationModule.kt

BenHenning committed:
Fix broken test builds.

There are still a bunch of test failures but they all relate to the
changes either on this branch or its parent.

BenHenning committed:
Merge branch 'fix-work-manager-and-app-init' into add-support-for-remote-platform-parameters

BenHenning committed:
Merge branch 'fix-work-manager-and-app-init' into add-support-for-remote-platform-parameters

BenHenning committed:
Lint fixes.

adhiamboperes commented on utility/src/main/java/org/oppia/android/util/properties/CustomPropertyRetrieverDebugImpl.kt:
```diff
@@ -0,0 +1,21 @@
+  private val context: Context
+) : CustomPropertyRetriever {
+  private val packageName = "org.oppia.android"
+  private val contentResolver by lazy { context.contentResolver }
```
Is there benefit to `contentResolver` being lazy delegated?

adhiamboperes reviewed (commented):
Thanks @BenHenning, I took a look at all the changes, but should probably wait until the PR is out of draft for the rest.

BenHenning commented:
Not stale.

BenHenning committed:
Fix remote profile syncing.

This improves debuggability and improves coroutine failure resilience,
though much of that will probably be removed from this change.
Otherwise, this properly fixes trying to sync parameters when offline
and fixes actual syncing so that it works.

BenHenning committed:
Merge branch 'fix-work-manager-and-app-init' into add-support-for-remote-platform-parameters

Conflicts:
	data/src/main/java/org/oppia/android/data/backends/gae/NetworkLoggingInterceptor.kt

BenHenning committed:
Some static check fixes.

BenHenning committed:
Add (partially generated) KDocs for retriever.

BenHenning committed:
Add more partially generated KDocs.
